### PR TITLE
Journal

### PR DESCRIPTION
**Pull request by justnero**
## Summary

This is a suggested draft for journal

## Checklist

- [ ] Writing style
- [ ] API design standards

### CODE REVIEW

File: openapi/openapi.yaml [hunk 1]
```diff
@@ -280,6 +280,10 @@ tags:
       customer pays more than they owe or returns a product. For
       information on the credit memo resource, see
       [Resources](https://www.rebilly.com/docs/dev-docs/concepts/#resources).
+  - name: Journal
+    description: |-
+      Use journal accounts, entries and records to provide track customer
+      generated revenue.
```
Comment (adamaltman) [inline openapi/openapi.yaml:286]:
```suggestion
```

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
+type: object
+required:
+  - type
+  - journalEntryId
+  - customerId
+  - invoiceId
+  - invoiceItemId
+  - debitAccount
+  - creditAccount
+properties:
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:10]: What about createdTime and updatedTime?

File: openapi/components/schemas/Common/CommonPlan.yaml [hunk 3]
```diff
@@ -73,6 +73,22 @@ properties:
     type: boolean
     description: Whether a plan has a trial without recurring instructions.
     readOnly: true
+  recognition:
```
[the remark this answers is not in the document]
Comment (justnero) [reply openapi/components/schemas/Common/CommonPlan.yaml:76]: We can add it to the product-level as well. Right now, the granularity used in QuickBooks integration is plan-level.

[the remark this answers is not in the document]
Comment (adamaltman) [reply openapi/components/schemas/Common/CommonPlan.yaml:76]: I think that needs to be reworked. The granularity of the tax-level is on the product-level. I think it should be consistent. I don't think it should be an "as well" but an only.

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    type: number
+    format: double
+  debitAccountId:
+    description: ID of the debit journal account.
+    nullable: true
+    allOf:
+      - $ref: ../ResourceId.yaml
+  creditAccountId:
+    description: ID of the credit journal account.
+    nullable: true
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:59]: Incorrect use of nullable (see #1106 )

File: openapi/components/schemas/Product.yaml [hunk 5]
```diff
@@ -12,6 +12,22 @@ allOf:
...
         example: '4010'
+      recognition:
+        type: object
+        properties:
+          enabled:
+            type: boolean
+            default: false
+          debitAccountId:
+            description: ID of the debit journal account.
+            nullable: true
```
Comment (adamaltman) [inline openapi/components/schemas/Product.yaml:23]: Incorrect use of nullable

File: openapi/components/schemas/Product.yaml [hunk 6]
```diff
@@ -12,6 +12,22 @@ allOf:
...
+            allOf:
+              - $ref: ./ResourceId.yaml
+          creditAccountId:
+            description: ID of the credit journal account.
+            nullable: true
```
Comment (adamaltman) [inline openapi/components/schemas/Product.yaml:28]: Incorrect use of nullable.

File: openapi/paths/journal-accounts.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      $ref: ../components/responses/Forbidden.yaml
+post:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Create a journal account
+  operationId: PostJournalAccount
+  x-sdk-operation-name: createAccount
+  description: |
```
Comment (adamaltman) [inline openapi/paths/journal-accounts.yaml:44]:
```suggestion
  description: |-
```

File: openapi/components/schemas/Journal/JournalRecord.yaml (hunk 4 above)
Comment (2e3s) [reply openapi/components/schemas/Journal/JournalRecord.yaml:59]: Isn't the type `string` here given by `allOf`? So if we want to specify that `null` is an acceptable type, we would need `nullable` anyway.

Comment (2e3s) [reply openapi/components/schemas/Journal/JournalRecord.yaml:59]: It's incorrect that those options look rather optional than nullable, but I don't agree with the particular argument.

Comment (adamaltman) [reply openapi/components/schemas/Journal/JournalRecord.yaml:59]: No, it is not given by `allOf`. From the OpenAPI 3.0 spec:

> A true value adds "null" to the allowed type specified by the type keyword, only if type is explicitly defined within the same Schema Object. Other Schema Object constraints retain their defined behavior, and therefore may disallow the use of null as a value. A false value leaves the specified or default type unmodified. The default value is false.

File: openapi/components/schemas/Journal/JournalEntry.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        format: date
+        example: 2022-09-01
+      endDate:
+        type: string
+        format: date
+        example: 2022-09-30
+  currency:
+    description: Currency of the journal record revenue.
+    allOf:
+      - $ref: ../CurrencyCode.yaml
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalEntry.yaml:25]: I'm confused. Is this only a revenue journal? What about other journal entries like credits and refunds, payments, and unearned revenue?

File: openapi/components/schemas/Journal/JournalEntry.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    type: string
+  references:
+    type: array
+    readOnly: true
+    items:
+      $ref: ../Integrations/IntegrationReference.yaml
+  createdTime:
+    description: Creation date/time.
+    allOf:
+      - $ref: ../ServerTimestamp.yaml
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalEntry.yaml:36]:
```suggestion
    $ref: ../CreatedTime.yaml
```

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      - $ref: ../ResourceId.yaml
+  type:
+    type: string
+    readOnly: true
+    enum:
+      - automated
+      - manual
+  estimatedAmount:
+    description: |-
+      Amount of revenue estimated to be recognized at the scheduling date.
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:41]:
```suggestion
      Amount of revenue estimated to be recognized at the schedule date.
```

File: openapi/paths/journal-entries.yaml [hunk 11]
```diff
@@ -0,0 +1,68 @@
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal entries
+  operationId: GetJournalEntryCollection
+  x-sdk-operation-name: getAllEntries
+  description: |
+    Retrieve a list of journal entries.
```
Comment (adamaltman) [inline openapi/paths/journal-entries.yaml:10]:
```suggestion
  description: Retrieves a list of journal entries.
```

File: openapi/paths/journal-entries.yaml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+post:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Create a journal entry
+  operationId: PostJournalEntry
+  x-sdk-operation-name: createEntry
+  description: |
+    Create a journal entry.
```
Comment (adamaltman) [inline openapi/paths/journal-entries.yaml:45]:
```suggestion
  description: Creates a journal entry.
```

File: openapi/paths/journal-entries@{id}.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal entry
+  operationId: GetJournalEntry
+  x-sdk-operation-name: getEntry
+  description: |
+    Retrieve a journal entry with a predefined identifier.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}.yaml:12]:
```suggestion
  description: Retrieves a journal entry with a predefined identifier.
```

File: openapi/paths/journal-entries@{id}.yaml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+patch:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Update a journal entry
+  operationId: PatchJournalEntry
+  x-sdk-operation-name: updateEntry
+  description: |
+    Update a journal entry.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}.yaml:35]:
```suggestion
  description: Updates a journal entry.
```

File: openapi/paths/journal-entries@{id}@records.yaml [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal records
+  operationId: GetJournalRecordCollection
+  x-sdk-operation-name: getAllRecords
+  description: |
+    Retrieve a list of journal records.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}@records.yaml:12]:
```suggestion
  description: Retrieves a list of journal records.
```

File: openapi/paths/journal-entries@{id}@records.yaml [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+post:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Create a journal record
+  operationId: PostJournalRecord
+  x-sdk-operation-name: createRecord
+  description: |
+    Create a journal record.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}@records.yaml:47]:
```suggestion
  description: Creates a journal record.
```

File: openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal record
+  operationId: GetJournalRecord
+  x-sdk-operation-name: getRecord
+  description: |
+    Retrieve a journal record with a predefined identifier.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml:13]:
```suggestion
  description: Retrieves a journal record with a specified identifier.
```

File: openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+patch:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Update a journal record
+  operationId: PatchJournalRecord
+  x-sdk-operation-name: updateRecord
+  description: |
+    Update a journal record.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml:36]:
```suggestion
  description: Updates a journal record.
```

File: openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+delete:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Delete a journal record
+  operationId: DeleteJournalRecord
+  x-sdk-operation-name: deleteRecord
+  description: |
+    Delete a journal record.
```
Comment (adamaltman) [inline openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml:68]:
```suggestion
  description: Deletes a journal record.
```

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    type: number
+    format: double
+  recognizedAmount:
+    description: Amount of revenue recognized at the journal period end.
+    nullable: true
+    type: number
+    format: double
+  debitAccountId:
+    description: ID of the debit journal account.
+    nullable: true
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:54]: Incorrect use of nullable here. How can this be nullable though?

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 21]
```diff
@@ -0,0 +1,81 @@
+type: object
+required:
+  - type
+  - journalEntryId
+  - customerId
+  - invoiceId
+  - invoiceItemId
+  - debitAccount
+  - creditAccount
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:9]: where are these properties defined?

File: openapi/components/schemas/Product.yaml [hunk 22]
```diff
@@ -12,6 +12,22 @@ allOf:
         description: Accounting code of the product.
         type: string
         example: '4010'
+      recognition:
+        type: object
```
Comment (adamaltman) [inline openapi/components/schemas/Product.yaml:16]:
```suggestion
        type: object
        nullable: true
```
How about making the object nullable?

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+      - $ref: ../ResourceId.yaml
+  creditAccountId:
+    description: ID of the credit journal account.
+    nullable: true
+    allOf:
+      - $ref: ../ResourceId.yaml
+  createdTime:
+    description: Creation date/time.
+    allOf:
+      - $ref: ../ServerTimestamp.yaml
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:65]:
```suggestion
    $ref: ../CreatedTime.yaml
```

File: openapi/paths/journal-accounts.yaml [hunk 24]
```diff
@@ -0,0 +1,68 @@
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal accounts
+  operationId: GetJournalAccountCollection
+  x-sdk-operation-name: getAllAccounts
+  description: |
+    Retrieve a list of journal accounts.
```
Comment (adamaltman) [inline openapi/paths/journal-accounts.yaml:10]:
```suggestion
  description: Retrieves a list of journal accounts.
```

File: openapi/paths/journal-accounts.yaml [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+post:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Create a journal account
+  operationId: PostJournalAccount
+  x-sdk-operation-name: createAccount
+  description: |
+    Create a journal account.
```
Comment (adamaltman) [inline openapi/paths/journal-accounts.yaml:45]:
```suggestion
  description: Creates a journal account.
```

File: openapi/paths/journal-accounts@{id}.yaml [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal account
+  operationId: GetJournalAccount
+  x-sdk-operation-name: getAccount
+  description: |
+    Retrieve a journal account with a predefined identifier.
```
Comment (adamaltman) [inline openapi/paths/journal-accounts@{id}.yaml:12]:
```suggestion
  description: Retrieves a journal account with a specified identifier.
```

File: openapi/paths/journal-accounts@{id}.yaml [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+patch:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Update a journal account
+  operationId: PatchJournalAccount
+  x-sdk-operation-name: updateAccount
+  description: |
+    Update a journal account.
```
Comment (adamaltman) [inline openapi/paths/journal-accounts@{id}.yaml:35]:
```suggestion
  description: Updates a journal account.
```

**Review by adamaltman — CHANGES_REQUESTED** [review]
Overall, I'm not sure about the name journal "record". Also, it doesn't seem correctly designed.

**Comment by 2e3s** [comment]
> Also, it doesn't seem correctly designed.

@adamaltman do you mean the design in general, or only what you've highlighted in your comments?

File: openapi/components/schemas/Journal/JournalRecord.yaml (hunk 21 above)
Comment (justnero) [reply openapi/components/schemas/Journal/JournalRecord.yaml:9]:
```suggestion
  - debitAccountId
  - creditAccountId
```

File: openapi/components/schemas/Journal/JournalEntry.yaml (hunk 8 above)
Comment (justnero) [reply openapi/components/schemas/Journal/JournalEntry.yaml:25]: They all follow the same debit/credit logic, right now automation is scoped to revenue only, but can be extended in future.

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+    $ref: ../CreatedTime.yaml
+  updatedTime:
+    $ref: ../UpdatedTime.yaml
+  _embedded:
+    type: array
+    description: |-
+      Embedded objects that are requested by the `expand`
+      query parameter.
+    readOnly: true
+    minItems: 1
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:62]:
```suggestion
    minItems: 0
```
?

Comment (2e3s) [reply openapi/components/schemas/Journal/JournalRecord.yaml:62]: I think it was copypasted, this constraint looks confusing for this property at all.

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+  creditAccountId:
+    description: ID of the credit journal account.
+    nullable: true
+    type: string
+  createdTime:
+    $ref: ../CreatedTime.yaml
+  updatedTime:
+    $ref: ../UpdatedTime.yaml
+  _embedded:
+    type: array
```
Comment (adamaltman) [inline openapi/components/schemas/Journal/JournalRecord.yaml:57]: Seems like this is actually an object (if it is similar to our other APIs).

**Review by adamaltman — DISMISSED** [review]
There is the problem with the _embedded object (it's not an array), but this is otherwise good.

File: openapi/components/parameters/journalRecordId.yaml [hunk 30]
```diff
@@ -0,0 +1,8 @@
+name: journalRecordId
+in: path
+description: Journal record identifier string.
```
Comment (noon-dawg) [inline openapi/components/parameters/journalRecordId.yaml:3]: nitpick: based on our style
```suggestion
description: ID of the journal record.
```

File: openapi/components/schemas/Journal/JournalAccount.yaml [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+      - $ref: ../ResourceId.yaml
+  name:
+    type: string
+    description: Name of the journal account.
+    example: Unearned revenue
+  description:
+    type: string
+    nullable: true
+  createdTime:
+    description: Creation date/time.
```
Comment (noon-dawg) [inline openapi/components/schemas/Journal/JournalAccount.yaml:18]: issue: [Timestamp fields](https://github.com/Rebilly/api-definitions/blob/main/WRITING-STYLE.md#timestamp-fields)
```suggestion
    description: Date and time when the journal record was created.
```

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+  type:
+    type: string
+    readOnly: true
+    enum:
+      - automated
+      - manual
+  estimatedAmount:
+    description: |-
+      Amount of revenue estimated to be recognized at the schedule date.
+      Value is ignored when updating a journal record with `type` of `automated`.
```
Comment (noon-dawg) [inline openapi/components/schemas/Journal/JournalRecord.yaml:35]: suggestion: suggestion for readability 
```suggestion
      This value is ignored when updating a journal record with a `type` of `automated`.
```

File: openapi/components/schemas/Journal/JournalRecord.yaml [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+        $ref: ../Invoices/Invoice.yaml
+      invoiceItem:
+        $ref: ../Invoices/InvoiceItem.yaml
+      debitAccount:
+        $ref: ./JournalAccount.yaml
+      creditAccount:
+        $ref: ./JournalAccount.yaml
+  _links:
+    type: array
+    description: Related resource links.
```
Comment (noon-dawg) [inline openapi/components/schemas/Journal/JournalRecord.yaml:73]: nitpick: based on our style
```suggestion
    description: Related links.
```

File: openapi/paths/journal-accounts@{id}.yaml [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+  - $ref: ../components/parameters/resourceId.yaml
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal account
+  operationId: GetJournalAccount
+  x-sdk-operation-name: getAccount
+  description: Retrieves a journal account with a predefined identifier.
```
Comment (noon-dawg) [inline openapi/paths/journal-accounts@{id}.yaml:11]: nitpick: based on our style
```suggestion
  description: Retrieves a journal account with a specified ID.
```

File: openapi/paths/journal-entries@{id}.yaml [hunk 35]
```diff
@@ -0,0 +1,64 @@
...
+  - $ref: ../components/parameters/resourceId.yaml
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal entry
+  operationId: GetJournalEntry
+  x-sdk-operation-name: getEntry
+  description: Retrieves a journal entry with a predefined identifier.
```
Comment (noon-dawg) [inline openapi/paths/journal-entries@{id}.yaml:11]: nitpick: based on our style
```suggestion
  description: Retrieves a journal entry with a specified ID.
```

File: openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml [hunk 36]
```diff
@@ -0,0 +1,85 @@
...
+  - $ref: ../components/parameters/journalRecordId.yaml
+get:
+  x-products:
+    - Core
+  tags:
+    - Journal
+  summary: Retrieve journal record
+  operationId: GetJournalRecord
+  x-sdk-operation-name: getRecord
+  description: Retrieves a journal record with a predefined identifier.
```
Comment (noon-dawg) [inline openapi/paths/journal-entries@{id}@records@{journalRecordId}.yaml:12]: nitpick: based on our style
```suggestion
  description: Retrieves a journal record with a specified ID.
```

**Review by noon-dawg — COMMENTED** [review]
Nice use of our writing style!
I added a few minor suggestions

File: openapi/components/schemas/Journal/JournalAccount.yaml (hunk 31 above)
Comment (2e3s) [reply openapi/components/schemas/Journal/JournalAccount.yaml]: Past time is prohibited :wink:

**Comment by 2e3s** [comment]
devs please review

File: openapi/components/schemas/Journal/JournalAccount.yaml (hunk 31 above)
Comment (noon-dawg) [reply openapi/components/schemas/Journal/JournalAccount.yaml]: Thanks for spotting this 🙏

**Comment by noon-dawg** [comment]
Looks good to me!

**Review by 2e3s — APPROVED** [review]
On @noon-dawg behalf

**Review by ashkarpetin — APPROVED** [review]

**Pull request merged**